libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt

```diff
@@ -0,0 +1,135 @@
+ *   override fun onCreate() {
+ *     MapboxNavigationApp
+ *       .setup(this)
+ *       .attach(this)
+ *   }
+ *   override fun onResume() {
+ *     MapboxNavigationApp
+ *       .registerObserver(locationObserver)
+ *   }
+ *   override fun onPause() {
+ *     MapboxNavigationApp
+ *       .unregisterObserver(locationObserver)
+ *   }
+ * }
+ * ```
+ *
+ * At any point of the application you register/unregister observers
+ * [MapboxNavigationApp.registerObserver]
+ * that can observe data coming from the [MapboxNavigation] object.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+object MapboxNavigationApp {
+
+    private val carAppLifecycleOwner = CarAppLifecycleOwner()
+    private val mapboxNavigationOwner = MapboxNavigationOwner()
+
+    /**
+     * Get the lifecycle of the car and app. It is started when either the car or
+     * app is in foreground. When both the car and app have been closed, the lifecycle
+     * is stopped. The lifecycle is never destroyed.
+     */
+    val lifecycleOwner: LifecycleOwner = carAppLifecycleOwner
+
+    /**
+     * Call [MapboxNavigationApp.setup] to provide the application with [NavigationOptions].
+     */
+    fun setup(navigationOptions: NavigationOptions) = apply {
+        mapboxNavigationOwner.setup(navigationOptions)
+    }
+
+    /**
+     * Detect when any Activity is in the foreground. This can be used with [disable]
+     * if you would like to enable or disable specific user experiences.
+     *
+     * Must be called after [setup].
+     */
+    fun attachAllActivities() = apply {
+        val application = mapboxNavigationOwner.navigationOptions.applicationContext as Application
+        carAppLifecycleOwner.attachAllActivities(application)
+        carAppLifecycleOwner.lifecycle.addObserver(mapboxNavigationOwner.carAppLifecycleObserver)
+    }
+
+    /**
+     * Optional function to disable [MapboxNavigation] from being created
+     * when the app is foregrounded. Make sure to call [setup] when you are ready to
+     * start using [MapboxNavigation] in the application.
+     */
+    fun disable() {
+        carAppLifecycleOwner.lifecycle.removeObserver(mapboxNavigationOwner.carAppLifecycleObserver)
+        mapboxNavigationOwner.disable()
+    }
+
+    /**
+     * Individually attach a lifecycle onto the [MapboxNavigationApp].
+     * If the app has been [setup], then this lifecycle will cause
+     * [MapboxNavigation] to be created.
+     *
+     * It is not required to [detach] every lifecycle. Because when
+     * the [Lifecycle] reaches the terminal [Lifecycle.State.DESTROYED] state
+     * it will be removed from the observers automatically.
+     */
+    fun attach(lifecycleOwner: LifecycleOwner) = apply {
+        carAppLifecycleOwner.attach(lifecycleOwner)
+    }
+
+    /**
+     * Individually detach lifecycles from [MapboxNavigationApp].
+     * When all LifecycleOwners have been detached, this will cause all
+     * [MapboxNavigationObserver.onDetached].
```

Comment:
This is a little bit inconvenient. If we wanted to use `MapboxNavigation` in an activity and add/remove listener in the `onAttached`/`onDetached` callbacks having another activity/fragment/viewmodel/AndroidAuto still working with `MapboxNavigation` would prevent us from getting `onDetached` callback and causing leaks. This means that we couldn't trust `onAttached`/`onDetached` lifecycle methods and still use local `onStart`/`onStop` instead, which defeats the purpose.

Reply:
Do you have a suggestion here?

The current idea is to separate the concerns. For example, your implementation of `MapboxNavigationObserver` does not know of the onStart, onStop events. So if you create a `RouteLineMbxObserver` it does not know of the lifecycle events.

But the Activity, is concerned with the lifecycle events. So your activity code looks like this (simple form) 
``` kotlin
MyActivity {
   onResume() { MapboxNavigationApp.register(routeLineMbxObserver) }
   onPause() { MapboxNavigationApp.unregister(routeLineMbxObserver) }
}

RouteLineMbxObserver {
  onAttach(mapboxNavigation) { renderRouteLine(mapboxNavigation) }
  onDetach(mapboxNavigation) { clearRouteLine(mapboxNavigation) }
}
```

And then can render the route line with arbitrary events

``` kotlin
MyRoutePreview {
   routeAvailable() { MapboxNavigationApp.register(routeLineMbxObserver) }
   userDoesntWantToSeeRoute() { MapboxNavigationApp.register(routeLineMbxObserver) }
}
```

Reply:
If there's another observer registered with `MapboxNavigationApp`, calling
```
onDestroy() { MapboxNavigationApp.detach(this) }
```
will not call
```
RouteLineMbxObserver#onDetach
```
defeating the purpose in my opinion, you will not have a clear understanding of when to unregistered your listeners from `MapboxNavigation` and you'll leak.

I think that the logic which calls `MapboxNavigationObserver#onDetach` shouldn't be tied to the main reference counter, but rather be invoked for each lifecycle owner separately when it's being detached or destroyed.

Only the `MapboxNavigation` instance should be tied to the main reference counter and destroyed when all lifecycles are finished.

Reply:
> will not call RouteLineMbxObserver#onDetach
> defeating the purpose in my opinion

I agree that, detaching should call onDetach. I'm pretty sure it does this, and we should be able to prove it with a unit test. Looking into this

Reply:
Excellent catch @LukasPaczos! Added a test and fixed it